src/Storages/MergeTree/LoadedMergeTreeDataPartInfoForReader.h

```diff
@@ -25,6 +25,8 @@ class LoadedMergeTreeDataPartInfoForReader final : public IMergeTreeDataPartInfo
 
     DataPartStoragePtr getDataPartStorage() const override { return data_part->getDataPartStoragePtr(); }
 
+    DataPartPtr getDataPart() const override { return data_part; }
```

Comment:
we cannot add a method getDataPart because the point of the class was to get rid of data part prt as some implementations of IMergeTreeDataPartInfo might not have it. See https://github.com/ClickHouse/ClickHouse/pull/41008. 
Let's pass IMergeTreeDataPartInfoForReader into MergeTreeMarksLoader instead of DataPartPtr, it will solve the issue as well.